flink-contrib/flink-siddhi/src/main/java/org/apache/flink/contrib/siddhi/operator/AbstractSiddhiOperator.java

```diff
@@ -0,0 +1,265 @@
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.operators.Output;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.MultiplexingStreamRecordSerializer;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.tasks.StreamTask;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.siddhi.core.ExecutionPlanRuntime;
+import org.wso2.siddhi.core.SiddhiManager;
+import org.wso2.siddhi.core.stream.input.InputHandler;
+import org.wso2.siddhi.query.api.definition.AbstractDefinition;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+public abstract class AbstractSiddhiOperator<IN, OUT> extends AbstractStreamOperator<OUT> implements OneInputStreamOperator<IN, OUT> {
+	private static final Logger LOGGER = LoggerFactory.getLogger(AbstractSiddhiOperator.class);
+	private static final int INITIAL_PRIORITY_QUEUE_CAPACITY = 11;
+
+	private final SiddhiOperatorContext siddhiPlan;
+	private final String executionExpression;
+	private final boolean isProcessingTime;
+	private final Map<String, MultiplexingStreamRecordSerializer<IN>> streamRecordSerializers;
+
+	private transient SiddhiManager siddhiManager;
+	private transient ExecutionPlanRuntime siddhiRuntime;
+	private transient Map<String, InputHandler> inputStreamHandlers;
+
+	// queue to buffer out of order stream records
+	private transient PriorityQueue<StreamRecord<IN>> priorityQueue;
+
+	/**
+	 * @param siddhiPlan Siddhi CEP  Execution Plan
+	 */
+	public AbstractSiddhiOperator(SiddhiOperatorContext siddhiPlan) {
+		validate(siddhiPlan);
+		this.executionExpression = siddhiPlan.getFinalExecutionPlan();
+		this.siddhiPlan = siddhiPlan;
+		this.isProcessingTime = this.siddhiPlan.getTimeCharacteristic() == TimeCharacteristic.ProcessingTime;
+		this.streamRecordSerializers = new HashMap<>();
+
+		for (String streamId : this.siddhiPlan.getInputStreams()) {
+			streamRecordSerializers.put(streamId, createStreamRecordSerializer(this.siddhiPlan.getInputStreamSchema(streamId), this.siddhiPlan.getExecutionConfig()));
+		}
+	}
+
+	protected abstract MultiplexingStreamRecordSerializer<IN> createStreamRecordSerializer(StreamSchema streamSchema, ExecutionConfig executionConfig);
+
+	protected MultiplexingStreamRecordSerializer<IN> getStreamRecordSerializer(String streamId) {
+		if (streamRecordSerializers.containsKey(streamId)) {
+			return streamRecordSerializers.get(streamId);
+		} else {
+			throw new UndefinedStreamException("Stream " + streamId + " not defined");
+		}
+	}
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
```

Comment:
This code seems to be similar to the code from CEP library. Can we reuse it somehow?